client/my-sites/themes/index.web.js

```diff
@@ -4,11 +4,20 @@
  * Internal dependencies
  */
 import userFactory from 'lib/user';
-import { makeLayout } from 'controller';
+import { makeLayout, redirectLoggedOut } from 'controller';
 import { navigation, siteSelection, sites } from 'my-sites/controller';
 import { loggedIn, loggedOut, upload, fetchThemeFilters } from './controller';
 import { validateFilters, validateVertical } from './validate-filters';
 
+function redirectToLoginIfSiteRequested( context, next ) {
+	if ( context.params.site_id ) {
+		redirectLoggedOut( context, next );
+		return;
+	}
+
+	next();
+}
+
```

Comment:
Bit of a style nit, but I kind of prefer to make route information as explicit as possible in route defs (i.e. `router()` statements, in this case) -- rather than having a check if there really _is_ a `site_id` param by this middleware.

How about:
1. Renaming `loggedOutRoutes` to `routesWithoutSiteFragments`
1. Renaming `loggedInRoutes` to `routesWithSiteFragments`, and removing the `?` after the `:site_id(${ siteId })`
1. In `if ( isLoggedIn ) { /*...*/ }`: Changing `router( routesWithSiteFragments, /*...*/ )` to `router( [ ...routesWithoutSiteFragments, ...routesWithSiteFragments ], /*...*/ )`
1. In the `else` branch: Add `router( routesWithSiteFragments, redirectLoggedOut );`

Reply:
Hi @ockham! This suggestion is very similar to my original implementation (visible [here](https://github.com/Automattic/wp-calypso/commit/a3c158b1b194ddfa66db1c5cc41f3621b18115dc), I force-pushed over that commit since then), but there is one gotcha:
```js
if ( loggedIn ) {
  router( [ ...routesWithoutSiteFragments, ...routesWithSiteFragments ] );
}
```
This doesn't work, because `/themes/site.blog` is matched by the `/themes/:vertical?` pattern rather than the `/themes/:site_id` one. Because of their order. Putting `routesWithSiteFragments` first should fix that. `/themes/site.blog` and `/themes/1234` will match the more restrictive `site_id` pattern, while `/themes/wedding` will fall through to the one with `vertical`.

It's solvable, but I'd rather land this PR first, as it works correctly, and then improve in a followup PR. There are folks blocked by this bug and waiting for a fix. And there is a risk that the improvements won't be so straightforward and will require a few review/testing cycles to make them right.